idaes/models/properties/modular_properties/phase_equil/smooth_VLE_2.py

```diff
@@ -0,0 +1,278 @@
+#################################################################################
+# The Institute for the Design of Advanced Energy Systems Integrated Platform
+# Framework (IDAES IP) was produced under the DOE Institute for the
+# Design of Advanced Energy Systems (IDAES).
+#
+# Copyright (c) 2018-2023 by the software owners: The Regents of the
+# University of California, through Lawrence Berkeley National Laboratory,
+# National Technology & Engineering Solutions of Sandia, LLC, Carnegie Mellon
+# University, West Virginia University Research Corporation, et al.
+# All rights reserved.  Please see the files COPYRIGHT.md and LICENSE.md
+# for full copyright and license information.
+#################################################################################
+"""
+Implementation of the formulation proposed in:
+
+Dabadghao, V., Ghouse, J., Eslick, J., Lee, A., Burgard, A., Miller, D., Biegler, L., 2022,
+A complementarity-based vapor-liquid equilibrium formulation for equation-oriented simulation
+and optimization. AIChE Journal, DOI: 10.1002/aic.18029
+"""
+
+from pyomo.environ import Constraint, Param, units as pyunits, Var, value
+
+from idaes.core.util.exceptions import ConfigurationError
+from idaes.core.util.math import smooth_min
+from idaes.models.properties.modular_properties.base.utility import (
+    identify_VL_component_list,
+)
+from idaes.models.properties.modular_properties.base.utility import (
+    estimate_Tbub,
+    estimate_Tdew,
+)
+import idaes.core.util.scaling as iscale
+
+
+# -----------------------------------------------------------------------------
+class SmoothVLE2:
+    """
+    Improved Vapor-Liquid Equilibrium complementarity formulation for Cubic Equations of State.
+    """
+
+    @staticmethod
+    def phase_equil(b, phase_pair):
+        # This method is called via StateBlock.build, thus does not need clean-up
+        # try/except statements
+        suffix = "_" + phase_pair[0] + "_" + phase_pair[1]
+
+        # Smooth VLE assumes a liquid and a vapor phase, so validate this
+        (
+            l_phase,
+            v_phase,
+            vl_comps,
+            henry_comps,
+            l_only_comps,
+            v_only_comps,
+        ) = identify_VL_component_list(b, phase_pair)
+
+        if l_phase is None or v_phase is None:
+            raise ConfigurationError(
+                f"{b.params.name} Generic Property Package phase pair {phase_pair[0]}-{phase_pair[1]} "
+                "was set to use Smooth VLE formulation, however this is not a vapor-liquid pair."
+            )
```

Comment:
This error should probably be thrown by `identify_VL_component_list`.

Reply:
A catch has been added to the utility method, but this can be retained as well.